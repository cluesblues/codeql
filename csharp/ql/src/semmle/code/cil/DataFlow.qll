/**
 * Provides a collection of building blocks and utilities for data flow.
 */

private import CIL

/**
 * A node in the data flow graph.
 *
 * Either an instruction (`Instruction`), a method return (`Method`), or a variable (`Variable`).
 */
class DataFlowNode extends @cil_dataflow_node {
  /** Gets a textual representation of this data flow node. */
  abstract string toString();

  /** Gets the type of this data flow node. */
  Type getType() { none() }

  /**
   * Holds if this node flows to `sink` in one step.
   * `tt` is the tainting that occurs during this step.
   */
  predicate getALocalFlowSucc(DataFlowNode sink, TaintType tt) {
    localExactStep(this, sink) and tt = TExactValue()
    or
    localTaintStep(this, sink) and tt = TTaintedValue()
  }

  private predicate flowsToStep(DataFlowNode sink) { this.getALocalFlowSucc(sink, TExactValue()) }

  /** Holds if this node flows to `sink` in zero or more steps. */
  predicate flowsTo(DataFlowNode sink) { this.flowsToStep*(sink) }

  /** Gets the method that contains this dataflow node. */
  Method getMethod() { none() }

  /** Gets the location of this dataflow node. */
  Location getLocation() { none() }
}

private newtype TTaintType =
  TExactValue() or
  TTaintedValue()

/** Describes how data is tainted. */
class TaintType extends TTaintType {
  string toString() {
    this = TExactValue() and result = "exact"
    or
    this = TTaintedValue() and result = "tainted"
  }
}

/** A taint type where the data is untainted. */
class Untainted extends TaintType, TExactValue { }

/** A taint type where the data is tainted. */
class Tainted extends TaintType, TTaintedValue { }

private predicate localExactStep(DataFlowNode src, DataFlowNode sink) {
  src = sink.(Opcodes::Dup).getAnOperand()
  or
  exists(VariableUpdate vu | src = vu.getSource() | variableUpdateAdjacentUse(_, vu, sink))
  or
  adjacentUseUse(src, sink)
  or
  src = sink.(ParameterReadAccess).getTarget()
  or
  src = sink.(Conversion).getExpr()
  or
  src = sink.(WriteAccess).getExpr()
  or
  src = sink.(Method).getAnImplementation().getAnInstruction().(Return)
  or
  src = sink.(Return).getExpr()
  or
  src = sink.(ConditionalBranch).getAnOperand()
  or
  src = sink.(Parameter).getAWrite()
  or
  exists(VariableUpdate update |
    update.getVariable().(Parameter) = sink and src = update.getSource()
  )
}

private predicate localTaintStep(DataFlowNode src, DataFlowNode sink) {
  src = sink.(BinaryArithmeticExpr).getAnOperand() or
  src = sink.(Opcodes::Neg).getOperand() or
  src = sink.(UnaryBitwiseOperation).getOperand()
}

module DefUse {
  /**
   * A classification of variable references into reads and writes.
   */
  private newtype RefKind =
    Read() or
    Write()

  /**
   * Holds if the `i`th node of basic block `bb` is a reference to `v`,
   * either a read (when `k` is `Read()`) or a write (when `k` is `Write()`).
   */
  private predicate ref(BasicBlock bb, int i, StackVariable v, RefKind k) {
    exists(ReadAccess ra | bb.getNode(i) = ra |
      ra.getTarget() = v and
      k = Read()
    )
    or
    exists(VariableUpdate vu | bb.getNode(i) = vu |
      vu.getVariable() = v and
      k = Write()
    )
  }

  /**
   * Gets the (1-based) rank of the reference to `v` at the `i`th node of
   * basic block `bb`, which has the given reference kind `k`.
   */
  private int refRank(BasicBlock bb, int i, StackVariable v, RefKind k) {
    i = rank[result](int j | ref(bb, j, v, _)) and
    ref(bb, i, v, k)
  }

  /** Gets the maximum rank index for the given variable and basic block. */
  private int lastRank(StackVariable v, BasicBlock bb) { result = max(refRank(bb, _, v, _)) }

  /**
   * Holds if stack variable `v` is live at the beginning of basic block `bb`.
   */
  private predicate liveAtEntry(BasicBlock bb, StackVariable v) {
    // The first reference to `v` inside `bb` is a read
    refRank(bb, _, v, Read()) = 1
    or
    // There is no reference to `v` inside `bb`, but `v` is live at entry
    // to a successor basic block of `bb`
    not ref(bb, _, v, _) and
    liveAtExit(bb, v)
  }

  /**
   * Holds if stack variable `v` is live at the end of basic block `bb`.
   */
  private predicate liveAtExit(BasicBlock bb, StackVariable v) {
    liveAtEntry(bb.getASuccessor(), v)
  }

  /**
   * Holds if the variable update `vu` reaches rank index `rankix`
   * in its own basic block `bb`.
   */
  private predicate defReachesRank(BasicBlock bb, VariableUpdate vu, int rankix, StackVariable v) {
    exists(int i |
      rankix = refRank(bb, i, v, Write()) and
      vu = bb.getNode(i)
    )
    or
    defReachesRank(bb, vu, rankix - 1, v) and
    rankix = refRank(bb, _, v, Read())
  }

  /** Holds if `v` is defined or used in `bb`. */
  private predicate varOccursInBlock(StackVariable v, BasicBlock bb) {
    exists(refRank(bb, _, v, _))
  }

  /** Holds if `v` occurs in `bb` or one of `bb`'s transitive successors. */
  private predicate blockPrecedesVar(StackVariable v, BasicBlock bb) {
    varOccursInBlock(v, bb)
    or
    liveAtExit(bb, v)
  }

  /**
   * Holds if `bb2` is a transitive successor of `bb1` and `v` occurs in `bb1` and
   * in `bb2` or one of its transitive successors but not in any block on the path
   * between `bb1` and `bb2`.
   */
  private predicate varBlockReaches(StackVariable v, BasicBlock bb1, BasicBlock bb2) {
    varOccursInBlock(v, bb1) and
    bb2 = bb1.getASuccessor() and
    blockPrecedesVar(v, bb2)
    or
    exists(BasicBlock mid |
      varBlockReaches(v, bb1, mid) and
      bb2 = mid.getASuccessor() and
      not varOccursInBlock(v, mid) and
      blockPrecedesVar(v, bb2)
    )
  }

  /**
   * Holds if `bb2` is a transitive successor of `bb1` and `v` occurs in `bb1` and
   * `bb2` but not in any block on the path between `bb1` and `bb2`.
   */
  private predicate varBlockStep(StackVariable v, BasicBlock bb1, BasicBlock bb2) {
    varBlockReaches(v, bb1, bb2) and
    varOccursInBlock(v, bb2)
  }

  /**
   * Holds if `v` occurs at index `i1` in `bb1` and at index `i2` in `bb2` and
   * there is a path between them without any occurrence of `v`.
   */
  private predicate adjacentVarRefs(StackVariable v, BasicBlock bb1, int i1, BasicBlock bb2, int i2) {
    exists(int rankix |
      bb1 = bb2 and
      rankix = refRank(bb1, i1, v, _) and
      rankix + 1 = refRank(bb2, i2, v, _)
    )
    or
    lastRank(v, bb1) = refRank(bb1, i1, v, _) and
    varBlockStep(v, bb1, bb2) and
    1 = refRank(bb2, i2, v, _)
  }

  cached
  private module Cached {
    /** Holds if the variable update `vu` can be used at the adjacent read `use`. */
    cached
    predicate variableUpdateAdjacentUse(StackVariable v, VariableUpdate vu, ReadAccess use) {
      exists(BasicBlock bb1, int i1, BasicBlock bb2, int i2 |
        adjacentVarRefs(v, bb1, i1, bb2, i2) and
        vu = bb1.getNode(i1) and
        use = bb2.getNode(i2)
      )
    }

    /** Holds if `use1` and `use2` are adjacent reads of the same variable. */
    cached
    predicate adjacentUseUse(ReadAccess use1, ReadAccess use2) {
      exists(BasicBlock bb1, int i1, BasicBlock bb2, int i2 |
        adjacentVarRefs(_, bb1, i1, bb2, i2) and
        use1 = bb1.getNode(i1) and
        use2 = bb2.getNode(i2)
      )
    }
  }

  import Cached

  /** Holds if the variable update `vu` can be used at the read `use`. */
  deprecated predicate variableUpdateUse(StackVariable target, VariableUpdate vu, ReadAccess use) {
    exists(ReadAccess ra |
      variableUpdateAdjacentUse(target, vu, ra) and
      adjacentUseUse*(ra, use)
    )
  }

  /** Holds if the update `def` can be used at the read `use`. */
  deprecated predicate defUse(StackVariable target, Expr def, ReadAccess use) {
    exists(VariableUpdate vu | def = vu.getSource() | variableUpdateUse(target, vu, use))
  }
}

private import DefUse

abstract library class VariableUpdate extends Instruction {
  abstract Expr getSource();

  abstract Variable getVariable();
}

private class VariableWrite extends VariableUpdate, WriteAccess {
  override Expr getSource() { result = getExpr() }

  override Variable getVariable() { result = getTarget() }
}

private class MethodOutOrRefTarget extends VariableUpdate, Call {
  int parameterIndex;

  MethodOutOrRefTarget() { this.getTarget().getRawParameter(parameterIndex).hasOutFlag() }

  override Variable getVariable() {
    result = this.getRawArgument(parameterIndex).(ReadAccess).getTarget()
  }

  override Expr getSource() { result = this }
}
