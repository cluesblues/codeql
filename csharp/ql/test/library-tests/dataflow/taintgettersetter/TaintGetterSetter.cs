public class A
{
    string Taint() { return "tainted"; }
    void Sink(object o) { }

    static string Step(string s) { return s + "0"; }

    class Box
    {
        public string s;
        public Box(string s)
        {
            this.s = s + "1";
        }
        public string GetS1() { return s + "2"; }
        public string GetS2() { return Step(s + "_") + "2"; }
        public void SetS1(string s) { this.s = "3" + s; }
        public void SetS2(string s) { this.s = "3" + Step("_" + s); }
        public static Box mk(string s)
        {
            Box b = new Box("");
            b.s = Step(s);
            return b;
        }
    }

    class Box2
    {
        string s1;
        string s2;
        public Box2(string s1, string s2)
        {
            this.s1 = s1 + "1";
            this.s2 = s2 + "2";
        }

        public string GetS1() { return s1 + "2"; }
        public string GetS2() { return Step(s2 + "_") + "2"; }
        public void SetS2(string s) { s2 = "3" + Step("_" + s); }
        public Box2 GetS1SetS2()
        {
            var b = new Box2("", "");
            b.SetS2(this.GetS1());
            return b;
        }
    }

    void M(Box b1, Box b2)
    {
        b1.SetS1(Taint());
        Sink(b1.GetS1());

        b2.SetS2(Taint());
        Sink(b2.GetS2());

        string t3 = Taint();
        var b3 = new Box(Step(t3));
        Sink(b3.s);

        var b4 = Box.mk(Taint());
        Sink(b4.GetS1());

        var b5 = new Box2(Taint(), "");
        var b6 = b5.GetS1SetS2();
        Sink(b6.GetS2());
    }
}
