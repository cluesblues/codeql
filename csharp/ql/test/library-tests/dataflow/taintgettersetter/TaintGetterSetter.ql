/**
 * @kind path-problem
 */

import csharp
import DataFlow::PathGraph

class Conf extends DataFlow::Configuration {
  Conf() { this = "taintgettersetter" }

  override predicate isSource(DataFlow::Node n) {
    n.asExpr().(MethodCall).getTarget().hasName("Taint")
  }

  override predicate isSink(DataFlow::Node n) {
    exists(MethodCall sink | sink.getAnArgument() = n.asExpr() and sink.getTarget().hasName("Sink"))
  }

  override predicate isAdditionalFlowStep(DataFlow::Node n1, DataFlow::Node n2) {
    exists(AddExpr add |
      add.getType() instanceof StringType and add.getAnOperand() = n1.asExpr() and n2.asExpr() = add
    )
  }
}

from DataFlow::PathNode source, DataFlow::PathNode sink, Conf conf
where conf.hasFlowPath(source, sink)
select sink, source, sink, "$@", source, source.toString()
