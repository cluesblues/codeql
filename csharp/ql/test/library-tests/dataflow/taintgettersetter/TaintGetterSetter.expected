edges
| TaintGetterSetter.cs:50:9:50:10 | [post] access to parameter b1 [s] : String | TaintGetterSetter.cs:51:14:51:15 | access to parameter b1 [s] : String |
| TaintGetterSetter.cs:50:18:50:24 | call to method Taint : String | TaintGetterSetter.cs:50:9:50:10 | [post] access to parameter b1 [s] : String |
| TaintGetterSetter.cs:51:14:51:15 | access to parameter b1 [s] : String | TaintGetterSetter.cs:51:14:51:23 | call to method GetS1 |
| TaintGetterSetter.cs:53:9:53:10 | [post] access to parameter b2 [s] : String | TaintGetterSetter.cs:54:14:54:15 | access to parameter b2 [s] : String |
| TaintGetterSetter.cs:53:18:53:24 | call to method Taint : String | TaintGetterSetter.cs:53:9:53:10 | [post] access to parameter b2 [s] : String |
| TaintGetterSetter.cs:54:14:54:15 | access to parameter b2 [s] : String | TaintGetterSetter.cs:54:14:54:23 | call to method GetS2 |
| TaintGetterSetter.cs:56:21:56:27 | call to method Taint : String | TaintGetterSetter.cs:57:31:57:32 | access to local variable t3 : String |
| TaintGetterSetter.cs:57:18:57:34 | object creation of type Box [s] : String | TaintGetterSetter.cs:58:14:58:15 | access to local variable b3 [s] : String |
| TaintGetterSetter.cs:57:26:57:33 | call to method Step : String | TaintGetterSetter.cs:57:18:57:34 | object creation of type Box [s] : String |
| TaintGetterSetter.cs:57:31:57:32 | access to local variable t3 : String | TaintGetterSetter.cs:57:26:57:33 | call to method Step : String |
| TaintGetterSetter.cs:58:14:58:15 | access to local variable b3 [s] : String | TaintGetterSetter.cs:58:14:58:17 | access to field s |
| TaintGetterSetter.cs:60:18:60:32 | call to method mk [s] : String | TaintGetterSetter.cs:61:14:61:15 | access to local variable b4 [s] : String |
| TaintGetterSetter.cs:60:25:60:31 | call to method Taint : String | TaintGetterSetter.cs:60:18:60:32 | call to method mk [s] : String |
| TaintGetterSetter.cs:61:14:61:15 | access to local variable b4 [s] : String | TaintGetterSetter.cs:61:14:61:23 | call to method GetS1 |
nodes
| TaintGetterSetter.cs:50:9:50:10 | [post] access to parameter b1 [s] : String | semmle.label | [post] access to parameter b1 [s] : String |
| TaintGetterSetter.cs:50:18:50:24 | call to method Taint : String | semmle.label | call to method Taint : String |
| TaintGetterSetter.cs:51:14:51:15 | access to parameter b1 [s] : String | semmle.label | access to parameter b1 [s] : String |
| TaintGetterSetter.cs:51:14:51:23 | call to method GetS1 | semmle.label | call to method GetS1 |
| TaintGetterSetter.cs:53:9:53:10 | [post] access to parameter b2 [s] : String | semmle.label | [post] access to parameter b2 [s] : String |
| TaintGetterSetter.cs:53:18:53:24 | call to method Taint : String | semmle.label | call to method Taint : String |
| TaintGetterSetter.cs:54:14:54:15 | access to parameter b2 [s] : String | semmle.label | access to parameter b2 [s] : String |
| TaintGetterSetter.cs:54:14:54:23 | call to method GetS2 | semmle.label | call to method GetS2 |
| TaintGetterSetter.cs:56:21:56:27 | call to method Taint : String | semmle.label | call to method Taint : String |
| TaintGetterSetter.cs:57:18:57:34 | object creation of type Box [s] : String | semmle.label | object creation of type Box [s] : String |
| TaintGetterSetter.cs:57:26:57:33 | call to method Step : String | semmle.label | call to method Step : String |
| TaintGetterSetter.cs:57:31:57:32 | access to local variable t3 : String | semmle.label | access to local variable t3 : String |
| TaintGetterSetter.cs:58:14:58:15 | access to local variable b3 [s] : String | semmle.label | access to local variable b3 [s] : String |
| TaintGetterSetter.cs:58:14:58:17 | access to field s | semmle.label | access to field s |
| TaintGetterSetter.cs:60:18:60:32 | call to method mk [s] : String | semmle.label | call to method mk [s] : String |
| TaintGetterSetter.cs:60:25:60:31 | call to method Taint : String | semmle.label | call to method Taint : String |
| TaintGetterSetter.cs:61:14:61:15 | access to local variable b4 [s] : String | semmle.label | access to local variable b4 [s] : String |
| TaintGetterSetter.cs:61:14:61:23 | call to method GetS1 | semmle.label | call to method GetS1 |
#select
| TaintGetterSetter.cs:51:14:51:23 | call to method GetS1 | TaintGetterSetter.cs:50:18:50:24 | call to method Taint : String | TaintGetterSetter.cs:51:14:51:23 | call to method GetS1 | $@ | TaintGetterSetter.cs:50:18:50:24 | call to method Taint : String | call to method Taint : String |
| TaintGetterSetter.cs:54:14:54:23 | call to method GetS2 | TaintGetterSetter.cs:53:18:53:24 | call to method Taint : String | TaintGetterSetter.cs:54:14:54:23 | call to method GetS2 | $@ | TaintGetterSetter.cs:53:18:53:24 | call to method Taint : String | call to method Taint : String |
| TaintGetterSetter.cs:58:14:58:17 | access to field s | TaintGetterSetter.cs:56:21:56:27 | call to method Taint : String | TaintGetterSetter.cs:58:14:58:17 | access to field s | $@ | TaintGetterSetter.cs:56:21:56:27 | call to method Taint : String | call to method Taint : String |
| TaintGetterSetter.cs:61:14:61:23 | call to method GetS1 | TaintGetterSetter.cs:60:25:60:31 | call to method Taint : String | TaintGetterSetter.cs:61:14:61:23 | call to method GetS1 | $@ | TaintGetterSetter.cs:60:25:60:31 | call to method Taint : String | call to method Taint : String |
